{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This file uses the KEGG API to make a request for Brite ID data and then stores briteIDs and their corresponding functional annotation groups, subgroups, and annotations to a dataframe that will be accessed later to create a hierarchical dataframe with gene expression values (in hierarchy.ipynb).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "url = 'http://rest.kegg.jp/get/br:br08902'\n",
    "request = requests.get(url, allow_redirects=True)\n",
    "open('KEGG_brite_groups.txt', 'wb').write(request.content)\n",
    "del request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(columns=['Functional Annotation Group','Functional Annotation Subgroup','keggBriteId','Functional Annotation'])\n",
    "\n",
    "f = open('KEGG_brite_groups.txt', 'r')\n",
    "lines = f.readlines()\n",
    "row_ind = 0\n",
    "col_ind = 0\n",
    "while(row_ind<len(lines)):\n",
    "    #get functional annotation group\n",
    "    if(lines[row_ind][0:1]=='A'): \n",
    "        A = lines[row_ind][1:-1]\n",
    "        row_ind+=1 # next line\n",
    "        \n",
    "        #get functional annotation subgroup\n",
    "        while(row_ind<len(lines) and lines[row_ind][0:1]!='A'):\n",
    "            if(lines[row_ind].split('  ')[0]=='B'):\n",
    "                B = lines[row_ind].split('  ')\n",
    "                B = B[1][:-1]\n",
    "                row_ind += 1 # next line\n",
    "                \n",
    "                #get briteID and the annotation itself and update dataframe with these values\n",
    "                while(row_ind<len(lines) and lines[row_ind].split('    ')[0]=='C'):\n",
    "                    C = lines[row_ind].split('    ')\n",
    "                    C = C[1][:-1]\n",
    "                    briteID =  C.split(\"  \")[0][3:]\n",
    "                    annotation = C.split(\"  \")[1]\n",
    "                    df.loc[col_ind] = [A, B, briteID, annotation]\n",
    "                    col_ind+=1\n",
    "                    row_ind += 1 \n",
    "            else: \n",
    "                row_ind+=1\n",
    "    else: \n",
    "        row_ind+=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional Annotation Group</th>\n",
       "      <th>Functional Annotation Subgroup</th>\n",
       "      <th>keggBriteId</th>\n",
       "      <th>Functional Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Pathway and Brite</td>\n",
       "      <td>Pathway maps</td>\n",
       "      <td>8901</td>\n",
       "      <td>KEGG pathway maps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Pathway and Brite</td>\n",
       "      <td>Brite files</td>\n",
       "      <td>8902</td>\n",
       "      <td>BRITE hierarchy files</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Pathway and Brite</td>\n",
       "      <td>Brite files</td>\n",
       "      <td>8904</td>\n",
       "      <td>BRITE table files</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Pathway and Brite</td>\n",
       "      <td>Networks</td>\n",
       "      <td>8907</td>\n",
       "      <td>KEGG networks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Genes and Proteins</td>\n",
       "      <td>Orthologs and modules</td>\n",
       "      <td>0001</td>\n",
       "      <td>KEGG Orthology (KO)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>108</td>\n",
       "      <td>Organisms and Cells</td>\n",
       "      <td>Organisms</td>\n",
       "      <td>8603</td>\n",
       "      <td>Plant classification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>109</td>\n",
       "      <td>Organisms and Cells</td>\n",
       "      <td>Organisms</td>\n",
       "      <td>8605</td>\n",
       "      <td>Plant pathogens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>110</td>\n",
       "      <td>Organisms and Cells</td>\n",
       "      <td>Organisms</td>\n",
       "      <td>8607</td>\n",
       "      <td>Endosymbionts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>111</td>\n",
       "      <td>Organisms and Cells</td>\n",
       "      <td>Cells, tissues, and organs</td>\n",
       "      <td>8501</td>\n",
       "      <td>Human organ systems</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>112</td>\n",
       "      <td>Organisms and Cells</td>\n",
       "      <td>Cells, tissues, and organs</td>\n",
       "      <td>8502</td>\n",
       "      <td>Human cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Functional Annotation Group Functional Annotation Subgroup keggBriteId  \\\n",
       "0             Pathway and Brite                   Pathway maps        8901   \n",
       "1             Pathway and Brite                    Brite files        8902   \n",
       "2             Pathway and Brite                    Brite files        8904   \n",
       "3             Pathway and Brite                       Networks        8907   \n",
       "4            Genes and Proteins          Orthologs and modules        0001   \n",
       "..                          ...                            ...         ...   \n",
       "108         Organisms and Cells                      Organisms        8603   \n",
       "109         Organisms and Cells                      Organisms        8605   \n",
       "110         Organisms and Cells                      Organisms        8607   \n",
       "111         Organisms and Cells     Cells, tissues, and organs        8501   \n",
       "112         Organisms and Cells     Cells, tissues, and organs        8502   \n",
       "\n",
       "     Functional Annotation  \n",
       "0        KEGG pathway maps  \n",
       "1    BRITE hierarchy files  \n",
       "2        BRITE table files  \n",
       "3            KEGG networks  \n",
       "4      KEGG Orthology (KO)  \n",
       "..                     ...  \n",
       "108   Plant classification  \n",
       "109        Plant pathogens  \n",
       "110          Endosymbionts  \n",
       "111    Human organ systems  \n",
       "112            Human cells  \n",
       "\n",
       "[113 rows x 4 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('KEGG_brite.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
